File: Sources/ClaudeWatcher.swift
// Watches Claude Code sessions by tailing the transcripts it writes under
// ~/.claude/projects/<project>/<session>.jsonl. Every line is a JSON object; the ones
// that matter here are human prompts, assistant tool calls, tool results (with errors)
// and end-of-turn assistant messages. No configuration is needed: every Claude Code
// front end (terminal, desktop app, IDE) writes the same files.
//
// Hooks (see `bitling claude`) deliver the same kinds faster and add permission
// notifications; the host de-duplicates when both are active.

import Foundation

final class ClaudeWatcher {
    var onEvent: ((GitEvent) -> Void)?

    private struct Session {
        let file: URL
        let sessionId: String
        var offset: UInt64
        var project: String
        var title: String
        var lastActivity: Date
        var lastToolEvent: Date
        var working: Bool
        var idleReported: Bool
    }

    private static let todayKey = "claudeToday"
    private let root = FileManager.default.homeDirectoryForCurrentUser.appendingPathComponent(".claude/projects")
    private let queue = DispatchQueue(label: "app.bitling.claude", qos: .utility)
    private var sessions: [String: Session] = [:]
    private var scanTimer: Timer?
    private var pollTimer: Timer?
    private var started = Date()

    private(set) var promptsToday = 0
    private(set) var toolsToday = 0
    private(set) var sessionsToday = 0
    private var todayKeyValue = ""
    private(set) var lastSummary = "nothing yet"

    var activeSessionCount: Int { queue.sync { sessions.values.filter { Date().timeIntervalSince($0.lastActivity) < 300 }.count } }

    init() {
        let defaults = UserDefaults.standard
        todayKeyValue = Self.dayString()
        if let today = defaults.dictionary(forKey: Self.todayKey), let day = today["day"] as? String, day == todayKeyValue {
            promptsToday = today["prompts"] as? Int ?? 0
            toolsToday = today["tools"] as? Int ?? 0
            sessionsToday = today["sessions"] as? Int ?? 0
        }
    }

    private static func dayString() -> String {
        let f = DateFormatter()
        f.dateFormat = "yyyy-MM-dd"
        return f.string(from: Date())
    }

    private func bump(prompts: Int = 0, tools: Int = 0, sessions: Int = 0) {
        let day = Self.dayString()
        if day != todayKeyValue { todayKeyValue = day; promptsToday = 0; toolsToday = 0; sessionsToday = 0 }
        promptsToday += prompts
        toolsToday += tools
        sessionsToday += sessions
        UserDefaults.standard.set(["day": day, "prompts": promptsToday, "tools": toolsToday, "sessions": sessionsToday], forKey: Self.todayKey)
    }

    func start() {
        started = Date()
        queue.async { self.scan() }
        scanTimer = Timer(timeInterval: 10.0, repeats: true) { [weak self] _ in self?.queue.async { self?.scan() } }
        pollTimer = Timer(timeInterval: 2.0, repeats: true) { [weak self] _ in self?.queue.async { self?.poll() } }
        for timer in [scanTimer!, pollTimer!] { RunLoop.main.add(timer, forMode: .common) }
    }

    // MARK: Discovery

    private func fileSize(_ url: URL) -> UInt64 {
        (try? FileManager.default.attributesOfItem(atPath: url.path)[.size] as? UInt64) ?? 0
    }

    private func scan() {
        let fm = FileManager.default
        guard let projects = try? fm.contentsOfDirectory(at: root, includingPropertiesForKeys: [.isDirectoryKey], options: [.skipsHiddenFiles]) else { return }
        let now = Date()
        for project in projects {
            guard (try? project.resourceValues(forKeys: [.isDirectoryKey]).isDirectory) == true else { continue }
            guard let files = try? fm.contentsOfDirectory(at: project, includingPropertiesForKeys: [.contentModificationDateKey, .creationDateKey], options: [.skipsHiddenFiles]) else { continue }
            for file in files where file.pathExtension == "jsonl" {
                let values = try? file.resourceValues(forKeys: [.contentModificationDateKey, .creationDateKey])
                let modified = values?.contentModificationDate ?? .distantPast
                guard now.timeIntervalSince(modified) < 15 * 60 else { continue }
                let key = file.path
                if sessions[key] != nil { continue }
                let created = values?.creationDate ?? modified
                let brandNew = created > started && now.timeIntervalSince(created) < 120
                let projectName = Self.projectName(fromFolder: project.lastPathComponent)
                sessions[key] = Session(
                    file: file, sessionId: file.deletingPathExtension().lastPathComponent,
                    offset: brandNew ? 0 : fileSize(file), project: projectName, title: "",
                    lastActivity: modified, lastToolEvent: .distantPast, working: false, idleReported: true
                )
                if brandNew {
                    bump(sessions: 1)
                    emit(GitEvent(kind: "claude-session-start", repo: projectName, branch: "", message: "", hash: "", name: projectName))
                }
            }
        }
        // Forget sessions that have been quiet for an hour.
        sessions = sessions.filter { now.timeIntervalSince($0.value.lastActivity) < 3600 }
    }

    private static func projectName(fromFolder folder: String) -> String {
        // "-Users-a12345-Desktop-AI-OyeChats" -> "OyeChats"
        let parts = folder.split(separator: "-").filter { !$0.isEmpty }
        return parts.last.map(String.init) ?? folder
    }

    // MARK: Tailing

    private func poll() {
        let now = Date()
        for (key, var session) in sessions {
            let size = fileSize(session.file)
            if size < session.offset { session.offset = 0 }
            if size > session.offset {
                let text = readTail(session.file, from: session.offset)
                session.offset = size
                session.lastActivity = now
                for line in text.split(separator: "\n") {
                    handle(line: String(line), session: &session)
                }
            } else if session.working, !session.idleReported, now.timeIntervalSince(session.lastActivity) > 5 * 60 {
                session.working = false
                session.idleReported = true
                emit(GitEvent(kind: "claude-idle", repo: session.project, branch: "", message: "", hash: "", name: session.project))
            }
            sessions[key] = session
        }
    }

    private func readTail(_ url: URL, from offset: UInt64) -> String {
        guard let handle = try? FileHandle(forReadingFrom: url) else { return "" }
        defer { try? handle.close() }
        do {
            try handle.seek(toOffset: offset)
            let data = try handle.readToEnd() ?? Data()
            return String(decoding: data, as: UTF8.self)
        } catch { return "" }
    }

    private static func preview(_ text: String, limit: Int = 60) -> String {
        let flat = text.replacingOccurrences(of: "\n", with: " ").trimmingCharacters(in: .whitespacesAndNewlines)
        return flat.count > limit ? String(flat.prefix(limit - 1)) + "…" : flat
    }

    private func handle(line: String, session: inout Session) {
        guard let data = line.data(using: .utf8),
              let entry = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let type = entry["type"] as? String else { return }
        if let cwd = entry["cwd"] as? String, !cwd.isEmpty { session.project = URL(fileURLWithPath: cwd).lastPathComponent }
        if entry["isSidechain"] as? Bool == true { return }
        let message = entry["message"] as? [String: Any]
        let content = message?["content"]

        switch type {
        case "custom-title":
            session.title = entry["customTitle"] as? String ?? session.title
        case "user":
            if entry["isMeta"] as? Bool == true { return }
            if let text = content as? String {
                guard (entry["origin"] as? [String: Any])?["kind"] as? String != "system" else { return }
                session.working = true
                session.idleReported = false
                bump(prompts: 1)
                emit(GitEvent(kind: "claude-prompt", repo: session.project, branch: "", message: Self.preview(text), hash: "", name: session.project))
            } else if let blocks = content as? [[String: Any]] {
                for block in blocks where block["type"] as? String == "tool_result" {
                    let failed = block["is_error"] as? Bool == true
                        || ((entry["toolUseResult"] as? [String: Any])?["success"] as? Bool == false)
                    if failed {
                        emit(GitEvent(kind: "claude-tool-error", repo: session.project, branch: "", message: "", hash: "", name: session.project))
                    }
                }
            }
        case "assistant":
            guard let blocks = content as? [[String: Any]] else { return }
            let stop = message?["stop_reason"] as? String ?? ""
            for block in blocks {
                let kind = block["type"] as? String ?? ""
                if kind == "tool_use" {
                    let tool = block["name"] as? String ?? "tool"
                    let input = block["input"] as? [String: Any] ?? [:]
                    let detail = (input["command"] as? String) ?? (input["file_path"] as? String) ?? (input["pattern"] as? String) ?? (input["query"] as? String) ?? ""
                    session.working = true
                    session.idleReported = false
                    bump(tools: 1)
                    if Date().timeIntervalSince(session.lastToolEvent) > 3 {
                        session.lastToolEvent = Date()
                        emit(GitEvent(kind: "claude-tool", repo: session.project, branch: "", message: Self.preview(detail, limit: 80), hash: "", name: tool))
                    }
                } else if kind == "text", stop == "end_turn" {
                    session.working = false
                    session.idleReported = true
                    emit(GitEvent(kind: "claude-done", repo: session.project, branch: "", message: Self.preview(block["text"] as? String ?? ""), hash: "", name: session.project))
                }
            }
        default:
            break
        }
    }

    private func emit(_ event: GitEvent) {
        lastSummary = "\(event.kind.replacingOccurrences(of: "claude-", with: "")) · \(event.name)"
        DispatchQueue.main.async { self.onEvent?(event) }
    }
}

File: apps/windows/Bitling.Windows/Watchers/ClaudeWatcher.cs
// Watches Claude Code sessions by tailing the transcripts it writes under
// %USERPROFILE%\.claude\projects\<project>\<session>.jsonl. Every line is a JSON object;
// the ones that matter here are human prompts, assistant tool calls, tool results (with
// errors) and end-of-turn assistant messages. No configuration is needed: every Claude
// Code front end (terminal, desktop app, IDE) writes the same files.
//
// Ported from apps/macos/Sources/ClaudeWatcher.swift.
using System.Text.Json.Nodes;
using Bitling.Models;
using Bitling.Settings;

namespace Bitling.Watchers;

public sealed class ClaudeWatcher
{
    public event Action<GitEvent>? OnEvent;

    private sealed class Session
    {
        public required string File;
        public required string SessionId;
        public long Offset;
        public string Project = "";
        public string Title = "";
        public DateTime LastActivity;
        public DateTime LastToolEvent = DateTime.MinValue;
        public bool Working;
        public bool IdleReported = true;
    }

    private const string TodayKey = "claudeToday";
    private readonly string _root = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".claude", "projects");
    private readonly Dictionary<string, Session> _sessions = new();
    private System.Threading.Timer? _scanTimer, _pollTimer;
    private DateTime _started = DateTime.Now;

    public int PromptsToday { get; private set; }
    public int ToolsToday { get; private set; }
    public int SessionsToday { get; private set; }
    private string _todayKeyValue = "";
    public string LastSummary { get; private set; } = "nothing yet";

    public int ActiveSessionCount { get { lock (_sessions) return _sessions.Values.Count(s => (DateTime.Now - s.LastActivity) < TimeSpan.FromMinutes(5)); } }

    private static string DayString() => DateTime.Now.ToString("yyyy-MM-dd");

    public ClaudeWatcher()
    {
        _todayKeyValue = DayString();
        var today = AppSettings.GetObject(TodayKey);
        if (today?["day"]?.GetValue<string>() == _todayKeyValue)
        {
            PromptsToday = today["prompts"]?.GetValue<int>() ?? 0;
            ToolsToday = today["tools"]?.GetValue<int>() ?? 0;
            SessionsToday = today["sessions"]?.GetValue<int>() ?? 0;
        }
    }

    private void Bump(int prompts = 0, int tools = 0, int sessions = 0)
    {
        var day = DayString();
        if (day != _todayKeyValue) { _todayKeyValue = day; PromptsToday = 0; ToolsToday = 0; SessionsToday = 0; }
        PromptsToday += prompts;
        ToolsToday += tools;
        SessionsToday += sessions;
        AppSettings.SetObject(TodayKey, new JsonObject { ["day"] = day, ["prompts"] = PromptsToday, ["tools"] = ToolsToday, ["sessions"] = SessionsToday });
    }

    public void Start()
    {
        _started = DateTime.Now;
        Task.Run(Scan);
        _scanTimer = new System.Threading.Timer(_ => Task.Run(Scan), null, TimeSpan.FromSeconds(10), TimeSpan.FromSeconds(10));
        _pollTimer = new System.Threading.Timer(_ => Task.Run(Poll), null, TimeSpan.FromSeconds(2), TimeSpan.FromSeconds(2));
    }

    private static long FileSize(string path) => File.Exists(path) ? new FileInfo(path).Length : 0;

    private void Scan()
    {
        if (!Directory.Exists(_root)) return;
        var now = DateTime.Now;
        IEnumerable<string> projects;
        try { projects = Directory.EnumerateDirectories(_root); } catch { return; }
        foreach (var project in projects)
        {
            IEnumerable<string> files;
            try { files = Directory.EnumerateFiles(project, "*.jsonl", SearchOption.AllDirectories); }
            catch { continue; }
            foreach (var file in files)
            {
                FileInfo info;
                try { info = new FileInfo(file); } catch { continue; }
                if ((now - info.LastWriteTime) >= TimeSpan.FromMinutes(15)) continue;
                lock (_sessions) { if (_sessions.ContainsKey(file)) continue; }
                var created = info.CreationTime;
                var brandNew = created > _started && (now - created) < TimeSpan.FromSeconds(120);
                var projectName = ProjectName(Path.GetFileName(project));
                var session = new Session
                {
                    File = file, SessionId = Path.GetFileNameWithoutExtension(file),
                    Offset = brandNew ? 0 : FileSize(file), Project = projectName, LastActivity = info.LastWriteTime,
                };
                lock (_sessions) _sessions[file] = session;
                if (brandNew)
                {
                    Bump(sessions: 1);
                    Emit(new GitEvent { Kind = "claude-session-start", Repo = projectName, Name = projectName });
                }
            }
        }
        // Forget sessions that have been quiet for an hour.
        lock (_sessions)
        {
            foreach (var key in _sessions.Where(kv => (now - kv.Value.LastActivity) >= TimeSpan.FromHours(1)).Select(kv => kv.Key).ToList())
                _sessions.Remove(key);
        }
    }

    private static string ProjectName(string folder)
    {
        // "-Users-a12345-Desktop-AI-OyeChats" -> "OyeChats"
        var parts = folder.Split('-', StringSplitOptions.RemoveEmptyEntries);
        return parts.Length > 0 ? parts[^1] : folder;
    }

    private void Poll()
    {
        var now = DateTime.Now;
        List<Session> snapshot;
        lock (_sessions) snapshot = _sessions.Values.ToList();
        foreach (var session in snapshot)
        {
            var size = FileSize(session.File);
            if (size < session.Offset) session.Offset = size;
            if (size > session.Offset)
            {
                var text = ReadTail(session.File, session.Offset);
                session.Offset = size;
                session.LastActivity = now;
                foreach (var line in text.Split('\n', StringSplitOptions.RemoveEmptyEntries))
                    Handle(line, session);
            }
            else if (session.Working && !session.IdleReported && (now - session.LastActivity) > TimeSpan.FromMinutes(5))
            {
                session.Working = false;
                session.IdleReported = true;
                Emit(new GitEvent { Kind = "claude-idle", Repo = session.Project, Name = session.Project });
            }
        }
    }

    private static string ReadTail(string path, long offset)
    {
        try
        {
            using var stream = new FileStream(path, FileMode.Open, FileAccess.Read, FileShare.ReadWrite);
            stream.Seek(offset, SeekOrigin.Begin);
            using var reader = new StreamReader(stream);
            return reader.ReadToEnd();
        }
        catch { return ""; }
    }

    private static string Preview(string text, int limit = 60)
    {
        var flat = text.Replace("\n", " ").Trim();
        return flat.Length > limit ? flat[..(limit - 1)] + "…" : flat;
    }

    private static DateTime? EntryDate(JsonNode? raw) =>
        raw is null ? null : (DateTime.TryParse(raw.GetValue<string>(), null,
            System.Globalization.DateTimeStyles.RoundtripKind, out var d) ? d : null);

    private void Handle(string line, Session session)
    {
        JsonObject? entry;
        try { entry = JsonNode.Parse(line) as JsonObject; } catch { return; }
        var type = entry?["type"]?.GetValue<string>();
        if (entry is null || type is null) return;
        var cwd = entry["cwd"]?.GetValue<string>();
        if (!string.IsNullOrEmpty(cwd)) session.Project = Path.GetFileName(cwd.TrimEnd('\\', '/'));
        if (entry["isSidechain"]?.GetValue<bool>() == true) return;
        if (EntryDate(entry["timestamp"]) is { } when && when < _started) return;
        var message = entry["message"] as JsonObject;
        var content = message?["content"];

        switch (type)
        {
            case "custom-title":
                session.Title = entry["customTitle"]?.GetValue<string>() ?? session.Title;
                break;
            case "user":
                if (entry["isMeta"]?.GetValue<bool>() == true) return;
                if (content is JsonValue textValue && textValue.TryGetValue<string>(out var text))
                {
                    if ((entry["origin"] as JsonObject)?["kind"]?.GetValue<string>() == "system") return;
                    session.Working = true;
                    session.IdleReported = false;
                    Bump(prompts: 1);
                    Emit(new GitEvent { Kind = "claude-prompt", Repo = session.Project, Message = Preview(text), Name = session.Project });
                }
                else if (content is JsonArray blocks)
                {
                    foreach (var block in blocks.OfType<JsonObject>().Where(b => b["type"]?.GetValue<string>() == "tool_result"))
                    {
                        var failed = block["is_error"]?.GetValue<bool>() == true
                            || (entry["toolUseResult"] as JsonObject)?["success"]?.GetValue<bool>() == false;
                        if (failed) Emit(new GitEvent { Kind = "claude-tool-error", Repo = session.Project, Name = session.Project });
                    }
                }
                break;
            case "assistant":
                if (content is not JsonArray abocks) return;
                var stop = message?["stop_reason"]?.GetValue<string>() ?? "";
                foreach (var block in abocks.OfType<JsonObject>())
                {
                    var kind = block["type"]?.GetValue<string>() ?? "";
                    if (kind == "tool_use")
                    {
                        var tool = block["name"]?.GetValue<string>() ?? "tool";
                        var input = block["input"] as JsonObject ?? new JsonObject();
                        var detail = input["command"]?.GetValue<string>() ?? input["file_path"]?.GetValue<string>()
                            ?? input["pattern"]?.GetValue<string>() ?? input["query"]?.GetValue<string>() ?? "";
                        session.Working = true;
                        session.IdleReported = false;
                        Bump(tools: 1);
                        if ((DateTime.Now - session.LastToolEvent) > TimeSpan.FromSeconds(3))
                        {
                            session.LastToolEvent = DateTime.Now;
                            Emit(new GitEvent { Kind = "claude-tool", Repo = session.Project, Message = Preview(detail, 80), Name = tool });
                        }
                    }
                    else if (kind == "text" && stop == "end_turn")
                    {
                        session.Working = false;
                        session.IdleReported = true;
                        Emit(new GitEvent { Kind = "claude-done", Repo = session.Project, Message = Preview(block["text"]?.GetValue<string>() ?? ""), Name = session.Project });
                    }
                }
                break;
        }
    }

    private void Emit(GitEvent e)
    {
        LastSummary = $"{e.Kind.Replace("claude-", "")} · {e.Name}";
        OnEvent?.Invoke(e);
    }
}
